src/object/traverse.ts

```diff
@@ -0,0 +1,199 @@
+
+    context.path.push(key)
+    const result = visitor(
+      (context.value = value),
+      (context.key = key),
+      context.parent,
+      context,
+    )
+
+    if (result === false) {
+      return (ok = false)
+    }
+
+    // Traverse nested plain objects and arrays that haven't been
+    // skipped and aren't a circular reference.
+    if (
+      value !== null &&
+      typeof value === 'object' &&
+      (isArray(value) || isPlainObject(value)) &&
+      !context.skipped.has(value) &&
+      !context.parents.includes(value)
+    ) {
+      traverse(value)
+
+      if (isFunction(result)) {
+        result()
+      }
+    }
+
+    context.path.pop()
+    return ok
+  }
+
+  const traverse = (parent: object, isRoot?: boolean): boolean => {
+    context.parents.push(parent)
+    context.parent = parent
+
+    if (isArray(parent)) {
+      // Use Array.prototype.forEach for arrays so that sparse arrays
+      // are efficiently traversed. The array must be cloned so it can
+      // be cleared if the visitor returns false.
+      parent.slice().forEach((value, index, values) => {
+        if (visit(value, index) === false) {
+          values.length = 0 // Stop further traversal.
+        }
+      })
+    }
+    // Allow an iterable (e.g. a Map or a Set) to be passed in as the
+    // root object.
+    else if (isRoot && isIterable(parent)) {
+      let index = 0
+      for (const value of parent) {
+        if (visit(value, index) === false) {
+          return false
+        }
+        index++
+      }
+    }
+    // Traverse the object's properties.
+    else {
+      for (const key of ownKeys(parent)) {
+        if (visit(parent[key as keyof object], key) === false) {
+          return false
+        }
+      }
+    }
+
+    context.parents.pop()
+    context.parent = last(context.parents)
+
+    return ok
+  }
+
+  // If this is a recursive call, it's possible that the root object
+  // was skipped earlier. Check for that here so the caller doesn't
+  // have to check for it.
+  return outerContext?.skipped.has(root) || traverse(root, true)
+}
+
+/**
+ * The visitor callback for the `traverse` function.
+ */
+export type TraverseVisitor = (
+  value: unknown,
+  key: keyof any,
+  parent: object,
+  context: TraverseContext,
+) => (() => void) | boolean | void
+
+/**
+ * The context object for the `traverse` function.
+ */
+export interface TraverseContext {
+  /**
+   * The current value being traversed.
+   */
+  readonly value: unknown
+  /**
+   * The property key or index with which the current value was found.
+   */
+  readonly key: keyof any
+  /**
+   * The parent object of the current value.
+   */
+  readonly parent: object
+  /**
+   * The stack of parent objects. The last object is the current
+   * parent.
+   *
+   * ⚠️ This array must not be mutated directly or referenced outside
+   * the `visitor` callback. If that's necessary, you'll want to clone
+   * it first.
+   */
+  readonly parents: readonly object[]
```

Comment:
I have a hacky solution for this problem. We can use getter for this `get parents` (or maybe method `getParents` would be even better (explicit)?). And we can copy array inside getter (so copying is lazy). To prevent a copy of stale array (outside `visitor` callback) we can keep `iterationIdx` in the context and call `iterationIdx++` each time we do recursive call. The code can look something like:

```js
const currentRunIdx = iterationIdx;
const context = {
    ...outerContext
    get parents() {
         if (currentRunIdx !== iterationIdx) {
              throw Error("traverse context was accessed outside of the callback")
         }

         return [...parents];
    }
}
``` 

The good thing is that we don't shadow the error here.  But with such API we always copy an array if user access it, which is not performant.  

In the end, I don't know how to solve it better.

Reply:
The getter isn't a bad idea, but it means we'd have to define another undocumented  property for the mutable array. I believe the current behavior is well-documented enough and misuse is noticeable enough, so it should be quick/easy for developers to debug their misuse without needing to complicate `traverse` further.